File: examples/lua/sleep.lua
-- hvlua coroutine sleep example: synchronous-style, non-blocking.
-- Run: bin/hvlua examples/lua/sleep.lua

hv.log("sleep example start")

-- This runs inside a coroutine, so hv.sleep() suspends WITHOUT blocking the
-- event loop: the two "workers" below interleave rather than run serially.

workers_done = 0

local function worker(name, ms)
    for i = 1, 3 do
        hv.log(name, "step", i)
        hv.sleep(ms)   -- looks blocking, actually yields to the loop
    end
    hv.log(name, "done")
    workers_done = workers_done + 1
    if workers_done == 2 then
        hv.stop()
    end
end

-- start two workers via timers so both run on the loop
-- (setTimeout requires timeout_ms >= 1; 0 is not a valid timer in libhv)
hv.setTimeout(1, function() worker("A", 300) end)
hv.setTimeout(1, function() worker("B", 500) end)

File: examples/js/sleep.js
// hvjs event-loop sleep example.
// Run: bin/hvjs examples/js/sleep.js

const hv = require("hv");

async function worker(name, ms) {
    for (let i = 1; i <= 3; ++i) {
        hv.log(name, "step", i);
        await hv.sleep(ms);
    }
    hv.log(name, "done");
}

await Promise.all([
    worker("A", 300),
    worker("B", 500),
]);

print("sleep example done");
